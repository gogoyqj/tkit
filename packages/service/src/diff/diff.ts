/**
 * @file: diff and patch
 * @author: yangqianjun
 * @Date: 2019-12-31 13:48:32
 * @LastEditors: yangqianjun
 * @LastEditTime: 2020-01-03 15:02:33
 */

// @IMP: f*ck
import * as jsondiffpatch from 'jsondiffpatch';
import lodash from 'lodash';

// @IMP: 确保浏览器和node使用相同的版本
// eslint-disable-next-line @typescript-eslint/no-var-requires
const realDiff: typeof jsondiffpatch = require('../../static/jsondiffpatch.umd');
const splitor = '__@__@__';

export function diffAndPatch(curVersion: {}, newVersion: {}) {
  const delta = realDiff.diff(curVersion, newVersion);
  return {
    delta,
    /** 按需更新 */
    selectDelta: (keys: string[][]) => {
      const initialMap: { [key: string]: any } = {};
      const setKeyMap: { [key: string]: any } = {};
      const newDelta = keys.reduce((newDelta, indexs) => {
        const val = lodash.get(delta, indexs);
        const newIndexs = indexs;
        const curSetKey = newIndexs.join(splitor);
        const parentIndexes = indexs.slice(0, indexs.length - 1);
        parentIndexes.reduce<string[]>((curKeys, curKey) => {
          curKeys.push(curKey);
          const key = curKeys.join(splitor);
          if (!(key in initialMap)) {
            const curValue = lodash.get(curVersion, curKeys);
            const newValue = lodash.get(newVersion, curKeys);
            if (
              (curValue && newValue && Array.isArray(curValue) && Array.isArray(newValue)) ||
              (curValue && !newValue && Array.isArray(curValue)) ||
              (!curValue && newValue && Array.isArray(newValue))
            ) {
              lodash.setWith(newDelta, curKeys.concat(['_t']), 'a', Object);
            }
            initialMap[key] = '';
          }
          return curKeys;
        }, []);
        if (!val) {
          throw `工具出 bug 了`;
        } else {
          lodash.setWith(newDelta, newIndexs, val, Object);
          setKeyMap[curSetKey] = '';
        }
        return newDelta;
      }, {});
      return newDelta;
    },
    /** 更新 */
    patch: (d: typeof delta) => {
      if (d) {
        realDiff.patch(curVersion, d);
        return curVersion;
      }
      return void 0;
    }
  };
}
