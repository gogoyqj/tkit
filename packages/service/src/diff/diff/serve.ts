/**
 * @file: serve a page for diff, select and patch
 * @author: yangqianjun
 * @Date: 2019-12-17 20:16:34
 * @LastEditors: yangqianjun
 * @LastEditTime: 2020-01-03 10:23:25
 */
/* eslint-disable @typescript-eslint/no-var-requires */

import detect from 'detect-port';
import express from 'express';
import bodyParser from 'body-parser';
import open from 'open';
import chalk from 'chalk';
import { StaticDir } from '../consts';
import { diffAndPatch } from './diff';

export async function serveDiff<J extends {}>(curVersion: J, newVersion: J) {
  return await detect(3007).then(port => {
    return new Promise<typeof curVersion | undefined>(resolve => {
      const { selectDelta, patch, delta } = diffAndPatch(curVersion, newVersion);
      const diffVersion = `${Math.random()}`;
      if (delta) {
        const app = express();
        let server: ReturnType<typeof app.listen> | undefined = undefined;
        const close = () => {
          server?.close();
        };
        app.use('/static', express.static(StaticDir, { cacheControl: true }));
        app.get('/', (req, res) => {
          res.send(`
<html>
  <head>
    <link rel="stylesheet" href="/static/html.css">
    <link rel="stylesheet" href="/static/annotated.css">
    <link rel="stylesheet" href="/static/service.css">
    <script src="/static/jsondiffpatch.umd.js"></script>
  </head>
  <body>
    <div id="menu" class="menu"></div>
    <div id="canvas" class="canvas"></div>
    <script>
      const diffVersion="${diffVersion}";
      const curVersion = ${JSON.stringify(curVersion, null, 2)};
      const delta = ${JSON.stringify(delta, null, 2)};
    </script>
    <script src="/static/service.js"></script>
  </body>
</html>
          `);
        });
        app.get('/diff', (req, res) => {
          res.json({
            code: 0,
            result: delta
          });
        });
        app.post('/patch', bodyParser.json(), (req, res) => {
          const { keys, version, unkeys } = req.body as {
            keys: string[][];
            unkeys: string[][];
            version: string;
          };
          if (version === diffVersion) {
            res.json({
              code: 0
            });
            // 全量
            if (!unkeys.length) {
              resolve(newVersion);
              // 增量
            } else if (keys.length) {
              const select = selectDelta(keys);
              patch(select);
              resolve(curVersion);
            }
            close();
          } else {
            res.json({
              code: 1,
              message: 'Diff 版本不一致，无法同步'
            });
          }
        });
        server = app.listen(port, () => {
          const url = `http://127.0.0.1:${port}/`;
          console.log(chalk.green(`[INFO]: 打开链接 ${url} 编辑`));
          open(url);
        });
        process.once('exit', close);
      } else {
        resolve();
      }
    });
  });
}
