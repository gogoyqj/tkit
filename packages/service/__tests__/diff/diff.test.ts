/**
 * @file: 确保 diff & patch 正确
 * @author: yangqianjun
 * @Date: 2020-01-03 11:40:45
 * @LastEditors: yangqianjun
 * @LastEditTime: 2020-01-03 15:29:36
 */

// @IMP: f*ck
import * as jsondiffpatch from 'jsondiffpatch';
import { diffAndPatch } from 'src/diff/diff';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const realDiff: typeof jsondiffpatch = require('../../static/jsondiffpatch.umd');

const samples = {
  arrStr: {
    p: ['a', 'b', 'c']
  },
  arrStrIncrease: {
    p: ['a', 'd', 'b', 'c', 'e']
  },
  arrStrDecrease: {
    p: ['c', 'd']
  },
  arrObj: {
    p: [
      {
        id: 1,
        name: 1
      },
      {
        id: 2,
        name: 2
      }
    ]
  },
  arrObjIncrease: {
    p: [
      {
        id: 1,
        name: 1
      },
      {
        id: 3,
        name: 3
      },
      {
        id: 2,
        name: 2
      }
    ]
  },
  arrObjDecrease: {
    p: [
      {
        id: 3,
        name: 3
      }
    ]
  },
  obj: {
    200: {
      id: 1,
      name: 1
    },
    500: {
      id: 2,
      name: 2
    }
  },
  objIncrease: {
    200: {
      id: 4,
      name: 4
    },
    404: {
      id: 3,
      name: 3
    },
    500: {
      id: 2,
      name: 2
    }
  },
  objDecrease: {
    500: {
      id: 3,
      name: 3
    }
  }
};
const getSmaples = (key: keyof typeof samples) => {
  return samples[key];
};
const clone = <T>(o: T): T => JSON.parse(JSON.stringify(o));

describe('service/diff ok', () => {
  it('array of string diffs ok', () => {
    const [Base, Inc, Dec] = [
      getSmaples('arrStr'),
      getSmaples('arrStrIncrease'),
      getSmaples('arrStrDecrease')
    ];
    // increase
    {
      const { delta, selectDelta } = diffAndPatch(Base, Inc);
      expect(delta).toMatchSnapshot('arrStrIncrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchObject(Inc);
      expect(realDiff.patch(clone(Base), delta!)).toMatchSnapshot('arrStrIncrease patched');

      {
        const newDelta = selectDelta([['p', '1']]);
        expect(newDelta).toMatchSnapshot('arrStrIncrease p-1');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot('arrStrIncrease p-1 patched');
      }

      {
        const newDelta = selectDelta([['p', '4']]);
        expect(newDelta).toMatchSnapshot('arrStrIncrease p-4');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot('arrStrIncrease p-4 patched');
      }
    }
    // decrease
    {
      const { delta, selectDelta } = diffAndPatch(Base, Dec);
      expect(delta).toMatchSnapshot('arrStrDecrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchSnapshot('arrStrDecrease patched');

      {
        const newDelta = selectDelta([['p', '1']]);
        expect(newDelta).toMatchSnapshot('arrStrDecrease p-1');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot('arrStrDecrease p-1 patched');
      }

      {
        const newDelta = selectDelta([
          ['p', '1'],
          ['p', '_1']
        ]);
        expect(newDelta).toMatchSnapshot('arrStrDecrease p-1&_1');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
          'arrStrDecrease p-1&_1 patched'
        );
      }
      {
        const newDelta = selectDelta([
          ['p', '_1'],
          ['p', '_0']
        ]);
        expect(newDelta).toMatchSnapshot('arrStrDecrease p-_1&_0');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
          'arrStrDecrease p-_1&_0 patched'
        );
      }
    }
  });

  it('array of obj diffs ok', () => {
    const [Base, Inc, Dec] = [
      getSmaples('arrObj'),
      getSmaples('arrObjIncrease'),
      getSmaples('arrObjDecrease')
    ];

    // increase
    {
      const { delta, selectDelta } = diffAndPatch(Base, Inc);
      expect(delta).toMatchSnapshot('arrObjIncrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchObject(Inc);
      expect(realDiff.patch(clone(Base), delta!)).toMatchSnapshot('arrObjIncrease patched');

      {
        {
          const newDelta = selectDelta([['p', '1', 'id']]);
          expect(newDelta).toMatchSnapshot('arrStrIncrease p-1-id');
          expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
            'arrObjIncrease p-1-id patched'
          );
        }

        {
          const newDelta = selectDelta([['p', '2']]);
          expect(newDelta).toMatchSnapshot('arrStrIncrease p-2');
          expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
            'arrObjIncrease p-2 patched'
          );
        }
      }
    }
    // decrease
    {
      const { delta, selectDelta } = diffAndPatch(Base, Dec);
      expect(delta).toMatchSnapshot('arrObjDecrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchObject(Dec);
      expect(realDiff.patch(clone(Base), delta!)).toMatchSnapshot('arrObjDecrease patched');

      {
        {
          const newDelta = selectDelta([['p', '0', 'name']]);
          expect(newDelta).toMatchSnapshot('arrObjDecrease p-0-name');
          expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
            'arrObjDecrease p-0-name patched'
          );
        }
        {
          const newDelta = selectDelta([['p', '_1']]);
          expect(newDelta).toMatchSnapshot('arrObjDecrease p-_1');
          expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
            'arrObjDecrease p-_1 patched'
          );
        }
      }
    }
  });

  it('object diffs ok', () => {
    const [Base, Inc, Dec] = [
      getSmaples('obj'),
      getSmaples('objIncrease'),
      getSmaples('objDecrease')
    ];

    // increase
    {
      const { delta, selectDelta } = diffAndPatch(Base, Inc);
      expect(delta).toMatchSnapshot('objIncrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchObject(Inc);

      {
        const newDelta = selectDelta([['200', 'id'], ['404']]);
        expect(newDelta).toMatchSnapshot('arrStrIncrease p-200&404');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
          'arrStrIncrease p-200&404 patched'
        );
      }

      {
        const newDelta = selectDelta([
          ['200', 'id'],
          ['200', 'name']
        ]);
        expect(newDelta).toMatchSnapshot('arrStrIncrease p-200');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
          'arrStrIncrease p-200 patched'
        );
      }
    }

    // decrease
    {
      const { delta, selectDelta } = diffAndPatch(Base, Dec);
      expect(delta).toMatchSnapshot('objDecrease');
      expect(realDiff.patch(clone(Base), delta!)).toMatchObject(Dec);

      {
        const newDelta = selectDelta([['200']]);
        expect(newDelta).toMatchSnapshot('objDecrease p-200');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot('objDecrease p-200 patched');
      }

      {
        const newDelta = selectDelta([['200'], ['500', 'name']]);
        expect(newDelta).toMatchSnapshot('objDecrease p-500-name');
        expect(realDiff.patch(clone(Base), newDelta)).toMatchSnapshot(
          'objDecrease p-500-name patched'
        );
      }
    }
  });
});
