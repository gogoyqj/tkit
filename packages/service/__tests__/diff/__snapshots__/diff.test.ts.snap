// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service/diff ok array of obj diffs ok: arrObjDecrease 1`] = `
Object {
  "p": Object {
    "0": Object {
      "id": Array [
        1,
        3,
      ],
      "name": Array [
        1,
        3,
      ],
    },
    "_1": Array [
      Object {
        "id": 2,
        "name": 2,
      },
      0,
      0,
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjDecrease p-_1 1`] = `
Object {
  "p": Object {
    "_1": Array [
      Object {
        "id": 2,
        "name": 2,
      },
      0,
      0,
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjDecrease p-_1 patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 1,
      "name": 1,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjDecrease p-0-name 1`] = `
Object {
  "p": Object {
    "0": Object {
      "name": Array [
        1,
        3,
      ],
    },
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjDecrease p-0-name patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 1,
      "name": 3,
    },
    Object {
      "id": 2,
      "name": 2,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjDecrease patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 3,
      "name": 3,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjIncrease 1`] = `
Object {
  "p": Object {
    "1": Object {
      "id": Array [
        2,
        3,
      ],
      "name": Array [
        2,
        3,
      ],
    },
    "2": Array [
      Object {
        "id": 2,
        "name": 2,
      },
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjIncrease p-1-id patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    Object {
      "id": 3,
      "name": 2,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjIncrease p-2 patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    Object {
      "id": 2,
      "name": 2,
    },
    Object {
      "id": 2,
      "name": 2,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrObjIncrease patched 1`] = `
Object {
  "p": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    Object {
      "id": 3,
      "name": 3,
    },
    Object {
      "id": 2,
      "name": 2,
    },
  ],
}
`;

exports[`service/diff ok array of obj diffs ok: arrStrIncrease p-1-id 1`] = `
Object {
  "p": Object {
    "1": Object {
      "id": Array [
        2,
        3,
      ],
    },
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of obj diffs ok: arrStrIncrease p-2 1`] = `
Object {
  "p": Object {
    "2": Array [
      Object {
        "id": 2,
        "name": 2,
      },
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease 1`] = `
Object {
  "p": Object {
    "1": Array [
      "d",
    ],
    "_0": Array [
      "a",
      0,
      0,
    ],
    "_1": Array [
      "b",
      0,
      0,
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-_1&_0 1`] = `
Object {
  "p": Object {
    "_0": Array [
      "a",
      0,
      0,
    ],
    "_1": Array [
      "b",
      0,
      0,
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-_1&_0 patched 1`] = `
Object {
  "p": Array [
    "c",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-1 1`] = `
Object {
  "p": Object {
    "1": Array [
      "d",
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-1 patched 1`] = `
Object {
  "p": Array [
    "a",
    "d",
    "b",
    "c",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-1&_1 1`] = `
Object {
  "p": Object {
    "1": Array [
      "d",
    ],
    "_1": Array [
      "b",
      0,
      0,
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease p-1&_1 patched 1`] = `
Object {
  "p": Array [
    "a",
    "d",
    "c",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrDecrease patched 1`] = `
Object {
  "p": Array [
    "c",
    "d",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease 1`] = `
Object {
  "p": Object {
    "1": Array [
      "d",
    ],
    "4": Array [
      "e",
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease p-1 1`] = `
Object {
  "p": Object {
    "1": Array [
      "d",
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease p-1 patched 1`] = `
Object {
  "p": Array [
    "a",
    "d",
    "b",
    "c",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease p-4 1`] = `
Object {
  "p": Object {
    "4": Array [
      "e",
    ],
    "_t": "a",
  },
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease p-4 patched 1`] = `
Object {
  "p": Array [
    "a",
    "b",
    "c",
    "e",
  ],
}
`;

exports[`service/diff ok array of string diffs ok: arrStrIncrease patched 1`] = `
Object {
  "p": Array [
    "a",
    "d",
    "b",
    "c",
    "e",
  ],
}
`;

exports[`service/diff ok object diffs ok: arrStrIncrease p-200 1`] = `
Object {
  "200": Object {
    "id": Array [
      1,
      4,
    ],
    "name": Array [
      1,
      4,
    ],
  },
}
`;

exports[`service/diff ok object diffs ok: arrStrIncrease p-200 patched 1`] = `
Object {
  "200": Object {
    "id": 4,
    "name": 4,
  },
  "500": Object {
    "id": 2,
    "name": 2,
  },
}
`;

exports[`service/diff ok object diffs ok: arrStrIncrease p-200&404 1`] = `
Object {
  "200": Object {
    "id": Array [
      1,
      4,
    ],
  },
  "404": Array [
    Object {
      "id": 3,
      "name": 3,
    },
  ],
}
`;

exports[`service/diff ok object diffs ok: arrStrIncrease p-200&404 patched 1`] = `
Object {
  "200": Object {
    "id": 4,
    "name": 1,
  },
  "404": Object {
    "id": 3,
    "name": 3,
  },
  "500": Object {
    "id": 2,
    "name": 2,
  },
}
`;

exports[`service/diff ok object diffs ok: objDecrease 1`] = `
Object {
  "200": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    0,
    0,
  ],
  "500": Object {
    "id": Array [
      2,
      3,
    ],
    "name": Array [
      2,
      3,
    ],
  },
}
`;

exports[`service/diff ok object diffs ok: objDecrease p-200 1`] = `
Object {
  "200": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    0,
    0,
  ],
}
`;

exports[`service/diff ok object diffs ok: objDecrease p-200 patched 1`] = `
Object {
  "500": Object {
    "id": 2,
    "name": 2,
  },
}
`;

exports[`service/diff ok object diffs ok: objDecrease p-500-name 1`] = `
Object {
  "200": Array [
    Object {
      "id": 1,
      "name": 1,
    },
    0,
    0,
  ],
  "500": Object {
    "name": Array [
      2,
      3,
    ],
  },
}
`;

exports[`service/diff ok object diffs ok: objDecrease p-500-name patched 1`] = `
Object {
  "500": Object {
    "id": 2,
    "name": 3,
  },
}
`;

exports[`service/diff ok object diffs ok: objIncrease 1`] = `
Object {
  "200": Object {
    "id": Array [
      1,
      4,
    ],
    "name": Array [
      1,
      4,
    ],
  },
  "404": Array [
    Object {
      "id": 3,
      "name": 3,
    },
  ],
}
`;
